src/main/cpp/faultinj/faultinj.cu

```diff
@@ -0,0 +1,506 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+#include <cuda.h>
+#include <cupti.h>
+#include <iostream>
+#include <map>
+#include <pthread.h>
+#include <exception>
+
+// thread-safe ptree
+#define BOOST_SPIRIT_THREADSAFE
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+
+#include <spdlog/spdlog.h>
+#include <sys/inotify.h>
+#include <sys/time.h>
+
+#define STDCALL
+
+extern "C" {
+
+#define CUPTI_CALL(call)                                                       \
+do {                                                                           \
+    CUptiResult _status = call;                                                \
+    if (_status != CUPTI_SUCCESS) {                                            \
+        const char *errstr;                                                    \
+        cuptiGetResultString(_status, &errstr);                                \
+        spdlog::error("function {} failed with error {}", #call, errstr);      \
+    }                                                                          \
+} while (0)
+
+#define PTHREAD_CALL(call)                                                         \
+do {                                                                               \
+    int _status = call;                                                            \
+    if (_status != 0) {                                                            \
+        spdlog::error("function {} failed with error code {}", #call, _status);    \
+    }                                                                              \
+} while (0)
+
+typedef enum {
+    FI_TRAP,
+    FI_ASSERT,
+    FI_RETURN_VALUE
+} FaultInjectionType;
+
+typedef struct {
+    volatile uint32_t initialized;
+    CUpti_SubscriberHandle  subscriber;
+
+    bool dynamic;
+    int terminateThread;
+    pthread_t dynamicThread;
+    // TODO change to the RAII idiom
+    pthread_rwlock_t configLock = PTHREAD_RWLOCK_INITIALIZER;
+
+    boost::property_tree::ptree configRoot;
+    boost::optional<boost::property_tree::ptree&> driverFaultConfigs = boost::none;
+    boost::optional<boost::property_tree::ptree&> runtimeFaultConfigs = boost::none;
+} injGlobalControl;
+injGlobalControl globalControl;
+
+// Function Declarations
+static void atExitHandler(void);
```

Comment:
Functions without attaching to any class can be just functions, don't need to have `static` keyword as they are already static.

Reply:
after we start using anonymous namespace which I'll try